### ISSUE TITLE
Add Shutdown Method to Exporter Interface and Existing Exporters

### ISSUE THREAD

**Issue by connorlindsey** [open-telemetry/opentelemetry-go#937]
The [Span Exporter Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#span-exporter) specifies that the Span Exporter must have an Export and Shutdown method.

Currently, there is no Exporter interface for Trace Exporters. Because there’s no interface, the existing Trace Exporters (Jaeger, Zipkin, Stdout) do not implement the Shutdown Method.

To resolve this issue, we should: 

- [ ] Add an interface for the TraceExporter which includes the Export and Shutdown methods defined in the spec linked above. 
- [ ] Refactor the existing Trace exporters to implement the new interface.

**Comment by lizthegrey** [comment]
@austinlparker @paulosman @MikeGoldsmith for awareness that we may need to change the Honeycomb & Lightstep exporters too.

**Comment by MrAlias** [comment]
I'm working on changes that use the following interface rework:

```diff
+// Exporter handles the delivery of processed SpanData to external
+// receivers. This is the final component in the trace export pipeline.
+type Exporter interface {
+       // Export handles the delivery of the passed SpanData batch.
+       //
+       // This function is called synchronously, so there is no concurrency
+       // safety requirement. However, due to the synchronous calling pattern,
+       // it is critical that all timeouts and cancellations contained in the
+       // passed context must be honored.
+       //
+       // Any retry logic must be contained in this function. The SDK that
+       // calls this function will not implement any retry logic. All errors
+       // returned by this function are considered unrecoverable and will be
+       // reported to a configured error Handler.
+       Export(context.Context, []*SpanData) error
+       // Shutdown notifies the exporter of a pending halt to operations. The
+       // exporter is expected to preform any cleanup or synchronization it
+       // requires while honoring all timeouts and cancellations contained in
+       // the passed context.
+       Shutdown(context.Context)
+}
-// SpanSyncer is a type for functions that receive a single sampled trace span.
-//
-// The ExportSpan method is called synchronously. Therefore, it should not take
-// forever to process the span.
-//
-// The SpanData should not be modified.
-type SpanSyncer interface {
-       ExportSpan(context.Context, *SpanData)
-}
- 
-// SpanBatcher is a type for functions that receive batched of sampled trace
-// spans.
-//
-// The ExportSpans method is called asynchronously. However its should not take
-// forever to process the spans.
-//
-// The SpanData should not be modified.
-type SpanBatcher interface {
-       ExportSpans(context.Context, []*SpanData)
-}
```

I went with a returned `error` from `Export` instead of a dedicated type as the specification alludes to (though has no normative statements requiring) because it serves the same purpose (`nil` is success, and not `nil` is an error).

I also included a context to `Shutdown` to handle timeout and cancelation so the SDK is not hung by an exporter and can communicate when things are over.

The interface name `Exporter` is the same as in the metrics package instead of mirroring the (again not normatively defined) specification.

Wondering if there are any thoughts on this?

**Comment by jmacd** [comment]
Is this a duplicate of #902?

**Comment by MrAlias** [comment]
> Is this a duplicate of #902?

I think #902 is related to the `SpanBatchProcessor` where this is related to the span exporter interface.

**Comment by MrAlias** [comment]
The `Exporter` method `Export` will clash with the similar metric `Exporter`. Will need to scope the method by telemetry type.

**Comment by MikeGoldsmith** [comment]
@MrAlias has there been any further work on introducing a `TraceExporter` interface? Is this something you plan finishing or would you like some help?

**Comment by MrAlias** [comment]
@MikeGoldsmith I'm still working on this. Each attempt has exposed other work that needed to be refactored first. With our big move in reduction to API dependencies this looks unblocked as is something I'm planning to get out this week. I'll be sure to update here with status when it changes.

**Comment by MikeGoldsmith** [comment]
Great, thanks for the update. I look forward to seeing your work 👍
